File: sources/iOS/ios-communications/iOSCommunications/ble/api/model/gatt/client/BleHrClient.swift

import Foundation
import CoreBluetooth
import RxSwift

public class BleHrClient: BleGattClientBase {
    
    public static let HR_SERVICE     = CBUUID(string: "180D")
    let BODY_SENSOR_LOCATION         = CBUUID(string: "2a38")
    public static let HR_MEASUREMENT = CBUUID(string: "2a37")
    public typealias BleHrNotification = (hr: Int, sensorContact: Bool, sensorContactSupported: Bool, energy: Int, rrs: [Int])
    var observers = AtomicList<RxObserver<BleHrNotification>>()
    
    public init(gattServiceTransmitter: BleAttributeTransportProtocol){
        super.init(serviceUuid: BleHrClient.HR_SERVICE, gattServiceTransmitter: gattServiceTransmitter)
        addCharacteristicNotification(BleHrClient.HR_MEASUREMENT)
        addCharacteristicRead(BODY_SENSOR_LOCATION)
    }
    
    // from base
    override public func disconnected() {
        super.disconnected()
        RxUtils.postErrorAndClearList(observers, error: BleGattException.gattDisconnected)
    }
    
    override public func processServiceData(_ chr: CBUUID , data: Data , err: Int ){
        if( chr.isEqual(BleHrClient.HR_MEASUREMENT) && err == 0 ){
            var offset=0
            let hrFormat = data[0] & 0x01
            let sensorContactBits = Int((data[0] & 0x06) >> 1)
            let energyExpended = (data[0] & 0x08) >> 3
            let rrPresent = (data[0] & 0x10) >> 4
            let sensorContact = sensorContactBits == 3
            let contactSupported = sensorContactBits != 0
            let hrValue = hrFormat == 1 ? (Int(data[1]) + (Int(data[2]) << 8)) : Int(data[1])
            offset = Int(hrFormat) + 2
            var energy = 0
            if (energyExpended == 1) {
                energy = Int(data[offset]) + (Int(data[offset + 1]) << 8)
                offset += 2
            }
            var rrs = [Int]()
            if( rrPresent == 1 ){
                let len = data.count
                while (offset < len) {
                    let rrValueRaw = Int(data[offset]) | (Int(data[offset + 1]) << 8)
                    offset += 2
                    rrs.append(rrValueRaw)
                }
            }
            RxUtils.emitNext(observers) { (observer) in
                observer.obs.onNext((hr: hrValue, sensorContact: sensorContact, sensorContactSupported: contactSupported,energy: energy,rrs: rrs))
            }
        }
    }
    
    public func observeHrNotifications(_ checkConnection: Bool) -> Observable<BleHrNotification> {
        return RxUtils.monitor(observers, transport: gattServiceTransmitter, checkConnection: checkConnection)
    }
    
    public override func clientReady(_ checkConnection: Bool) -> Completable {
        return waitNotificationEnabled(BleHrClient.HR_MEASUREMENT, checkConnection: checkConnection)
    }
}

File: sources/Android/android-communications/src/main/java/com/androidcommunications/polar/api/ble/model/gatt/client/BleHrClient.java
package com.androidcommunications.polar.api.ble.model.gatt.client;

import androidx.annotation.NonNull;

import com.androidcommunications.polar.api.ble.BleLogger;
import com.androidcommunications.polar.api.ble.model.gatt.BleGattBase;
import com.androidcommunications.polar.api.ble.model.gatt.BleGattTxInterface;
import com.androidcommunications.polar.common.ble.AtomicSet;
import com.androidcommunications.polar.common.ble.RxUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.FlowableEmitter;

public class BleHrClient extends BleGattBase {
    private static final String TAG = BleHrClient.class.getSimpleName();

    public static class HrNotificationData {
        public final int hrValue;
        public final boolean sensorContact;
        public final boolean sensorContactSupported;
        public final boolean rrPresent;
        public final int energy;
        public final long cumulativeRR;
        public final List<Integer> rrs = new ArrayList<>();

        public HrNotificationData(int hrValue,
                                  boolean sensorContact,
                                  int energy,
                                  List<Integer> rrs,
                                  boolean sensorContactSupported,
                                  long cumulativeRR,
                                  boolean rrPresent) {
            this.hrValue = hrValue;
            this.sensorContact = sensorContact;
            this.energy = energy;
            this.rrs.addAll(rrs);
            this.cumulativeRR = cumulativeRR;
            this.sensorContactSupported = sensorContactSupported;
            this.rrPresent = rrPresent;
        }
    }

    private final AtomicSet<FlowableEmitter<? super HrNotificationData>> hrObserverAtomicList = new AtomicSet<>();

    public static final UUID BODY_SENSOR_LOCATION = UUID.fromString("00002a38-0000-1000-8000-00805f9b34fb");
    public static final UUID HR_MEASUREMENT = UUID.fromString("00002a37-0000-1000-8000-00805f9b34fb");
    public static final UUID HR_SERVICE = UUID.fromString("0000180D-0000-1000-8000-00805f9b34fb");

    public BleHrClient(BleGattTxInterface txInterface) {
        super(txInterface, HR_SERVICE);
        addCharacteristicNotification(HR_MEASUREMENT);
        addCharacteristicRead(BODY_SENSOR_LOCATION);
    }

    @Override
    public void reset() {
        super.reset();
        RxUtils.postDisconnectedAndClearList(hrObserverAtomicList);
    }

    @Override
    public void processServiceData(UUID characteristic, byte[] data, int status, boolean notifying) {
        BleLogger.d(TAG, "Processing service data. Status: " + status + ".  Data length: " + data.length);
        if (status == 0 && characteristic.equals(HR_MEASUREMENT)) {
            long cumulativeRR = 0;
            final int hrFormat = data[0] & 0x01;
            final boolean sensorContact = ((data[0] & 0x06) >> 1) == 0x03;
            final boolean contactSupported = (data[0] & 0x04) != 0;
            int energyExpended = (data[0] & 0x08) >> 3;
            final int rrPresent = (data[0] & 0x10) >> 4;
            final int hrValue = (hrFormat == 1 ? (data[1] & 0xFF) + (data[2] << 8) : data[1]) & (hrFormat == 1 ? 0x0000FFFF : 0x000000FF);
            int offset = hrFormat + 2;
            int energy = 0;
            if (energyExpended == 1) {
                energy = (data[offset] & 0xFF) + ((data[offset + 1] & 0xFF) << 8);
                offset += 2;
            }
            final ArrayList<Integer> rrs = new ArrayList<>();
            if (rrPresent == 1) {
                int len = data.length;
                while (offset < len) {
                    int rrValue = ((data[offset] & 0xFF) + ((data[offset + 1] & 0xFF) << 8));
                    offset += 2;
                    rrs.add(rrValue);
                }
            }

            final long finalCumulativeRR = cumulativeRR;
            final int finalEnergy = energy;
            RxUtils.emitNext(hrObserverAtomicList, object ->
                    object.onNext(new HrNotificationData(hrValue, sensorContact, finalEnergy, rrs, contactSupported, finalCumulativeRR, rrPresent == 1)));
        }
    }

    @Override
    public void processServiceDataWritten(UUID characteristic, int status) {
        BleLogger.d(TAG, "Service data written not processed in BleHrClient");
    }

    @NonNull
    @Override
    public String toString() {
        // and so on
        return "HR gatt client";
    }

    @Override
    public Completable clientReady(boolean checkConnection) {
        return waitNotificationEnabled(HR_MEASUREMENT, checkConnection);
    }

    // API

    /**
     * @return Flowable stream
     * Produces: onNext, for every hr notification event
     * onError, if client is not initially connected or ble disconnect's
     * onCompleted, none except further configuration applied. If binded to fragment or activity life cycle this might be produced
     */
    public Flowable<HrNotificationData> observeHrNotifications(final boolean checkConnection) {
        return RxUtils.monitorNotifications(hrObserverAtomicList, txInterface, checkConnection);
    }
}
